### ISSUE TITLE
clush: Interactive usage - sudo with password and requiretty

### ISSUE THREAD

**Issue by hansfn** [cea-hpc/clustershell#234]
Hi,

I have used [polysh](http://guichaz.free.fr/polysh/) for several years while maintaining a small data center with CentOS servers. Today I switched to clustershell/clush, since I just started maintaining a tiny data center with Debian servers and polysh isn't packaged by Debian. clush seems to be a very good replacement, but I struggle with interactive usage and sudo with password when requiretty is enabled.

If I use

```
clush -o '-t -t' -w node[1-10]
```

the normal "sudo: no tty present and no askpass program specified" error is gone when I run "sudo date", but the password isn't captured if I enter it (and the password is displayed).

If I on the same setup use polysh

```
./polysh.py 'node<1-10>'
```

there is no problem with entering the password, and there is even a command ":hide_password" that stops the password from displaying.

Is there some way clush can handle this situation? If not, consider this a feature request. Maybe the code in polysh can be reused?

Added: Polysh 0.4, Clustershell 1.6

**Comment by thiell** [comment]
Sorry, didn't have the chance to work on that - postponed until 1.8 sorry

**Comment by underted** [comment]
Any update on adding sudo support? Could really use it. Our systems now require sudo and I've got to switch to using Ansible for adhoc commands and it isn't a great experience.

**Comment by martinetd** [comment]
I wrote a small hack ages ago that disables normal stdin (which is in another thread so messes main inuput) and prompts for password, just pushed here: https://github.com/martinetd/clustershell/commits/sudo

But it's really nothing more than a hack, and I personally stopped using sudo shortly after writing it so I'm not sure it still works (haven't tested in almost two years)

The stdin thread actually causes problems with python3 (#461), so it might be a good time to get rid of it but iirc there were problems with inserting stdin in the polling engine when stdin is a pipe for example? or something I don't recall, I wrote that ages ago.
If stdin is no longer in a separate thread, then it shouldn't be too difficult to prompt for a sudo password if a sudo prompt is detected for example.


The best would really to not rely on a password prompt though, even if clustershell works around that by forwarding a prompt that'll require disabling requiretty and leaves room for password hijacking. Best practice would be to have a sudo kerberos ticket forwarded and used for sudo for example, I think that should be possible, or some form of non-interactive authentication method...
